src/cmd/ksh93/bltins/read.c

```diff
@@ -127,7 +127,11 @@ int	b_read(int argc,char *argv[], Shbltin_t *context)
 		flags |= SS_FLAG;
 		break;
 	    case 'u':
+		if(opt_info.num<0 || opt_info.num>INT_MAX)
+			errormsg(SH_DICT,ERROR_exit(1),e_file,opt_info.arg); /* reject negative and out of range numbers */
 		fd = (int)opt_info.num;
+		if(fd>=shp->gd->lim.open_max && !sh_iovalidfd(shp,fd))
```

Comment:
Is there any reason not to test for both conditions here, and save compiling in an `errormsg` call? Like:

```c
	if(fd < 0 || fd > INT_MAX || (fd>=shp->gd->lim.open_max && !sh_iovalidfd(shp,fd)))
```

Reply:
The pull request has been updated with the optimization, although the first two tests must compare against `opt_info.num` as `fd` is only an `int`, which is too small for handling numbers outside of the 32-bit limit.

Reply:
Of course, good point, thanks.